xdsl/transforms/canonicalization_patterns/riscv.py

```diff
@@ -517,3 +542,18 @@ def get_constant_value(value: SSAValue) -> riscv.Imm32Attr | None:
 
     if isinstance(value.op, riscv.LiOp) and isinstance(value.op.immediate, IntegerAttr):
         return value.op.immediate
+
+
+def exact_log2(value: SSAValue) -> int | None:
+    if (v := get_constant_value(value)) is None:
+        return
+
+    if v.value.data < 1:
+        return
+
+    # Since we use 32-bit signed integers, we can hardcode all 31 powers of 2.
+    powers_of_two = [1 << i for i in range(31)]
+    try:
+        return powers_of_two.index(v.value.data)
+    except ValueError:
+        return
```

Comment:
```suggestion
if int.bit_count(v.value.data) == 1:
    return int.bit_length(v.value.data) - 1
```
Alternative to the list indexing

Reply:
That's better indeed. I was not aware Python had those methods. I still need to check that the value is positive though.